### ISSUE TITLE
[Improvement] Cleanup unused shuffle servers after stage completed

### ISSUE THREAD

**Issue by xianjingfeng** [apache/incubator-uniffle#328]
### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### Search before asking

- [X] I have searched in the [issues](https://github.com/apache/incubator-uniffle/issues?q=is%3Aissue) and found no similar issues.

### What would you like to be improved?

If there are many stages in one applicaiton, spark driver will send heartbeat to every shuffle servers, and it may cause app expired in shuffle server side. And if we support decommission in the future, it will cause shuffle difficult to exit.

### How should we improve?

Cleanup unused shuffle servers after stage completed

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

**Comment by jerqi** [comment]
Spark has a mechanism called `ExchangeReuse`. It means that we can use the older stage' shuffle data.

**Comment by jerqi** [comment]
Have you joined our `Wechat` group?

**Comment by xianjingfeng** [comment]
> Spark has a mechanism called `ExchangeReuse`. It means that we can use the older stage' shuffle data. We have already a pr #249, I don't know whether this pr can solve your problem.

#249 will clean the older stage' shuffle data, and it will not clean shuffle servers in shuffleServerInfoSet, right?

**Comment by xianjingfeng** [comment]
> Have you joined our `Wechat` group?

No, how to join it?

**Comment by jerqi** [comment]
> > Have you joined our `Wechat` group?
> 
> No, how to join it?

You can scan the QR code
![bcde7855d24455ec870a45a3f70df3c0](https://user-images.githubusercontent.com/8159038/202109313-1b2f86c5-35a4-4ed2-8122-066315b812e0.png)

**Comment by xianjingfeng** [comment]
> > > Spark has a mechanism called `ExchangeReuse`. It means that we can use the older stage' shuffle data. We have already a pr #249, I don't know whether this pr can solve your problem.
> > 
> > 
> > #249 will clean the older stage' shuffle data, and it will not clean shuffle servers in shuffleServerInfoSet, right?
> 
> Yes. you're right.

So it is ok if we cleanup unused shuffle servers?

**Comment by jerqi** [comment]
> > > > Spark has a mechanism called `ExchangeReuse`. It means that we can use the older stage' shuffle data. We have already a pr #249, I don't know whether this pr can solve your problem.
> > > 
> > > 
> > > #249 will clean the older stage' shuffle data, and it will not clean shuffle servers in shuffleServerInfoSet, right?
> > 
> > 
> > Yes. you're right.
> 
> So it is ok if we cleanup unused shuffle servers?

Yes, if we already removed the shuffle data in those shuffle servers, we can remove the shuffle servers in the shuffleServerInfoSet.
